### PR TITLE
LLDP: add API to export neighbors list (v4)

### PR DESCRIPTION
**Pull request by bengal**
v4 of #765

The LLDP support in libsystemd-network at the moment only allows to
save the list of discovered neighbors to a text file.

It would be useful to have an API to export information about
neighbors in a more flexible way to users of the library. These
patches export the TLV structure as a public opaque type and provide
public functions to access its content.

Changes since v3:
- don't throw assertion when dest MAC address is unknown
- added new function lldp_tlv_packet_enter_container_org()
- check return value of lldp_tlv_packet_exit_container()
- coding style

### CODE REVIEW

**Comment by bengal** [comment]
> The packets should be dropped as soon as the TTL expires. So if some one is holding those packets
> it will never get freed and it will look like the neighbour is still reachable if in case the it goes down.
> 
> Again since we only store number of packets is limited we will not receive new packets if this limit
> is reached even if TTL is expired.

The patch doesn't change the packet processing and expiry logic; when
the TTL for a frame is over, the entry is still removed from the MIB.

The only difference introduced by the patch is that if someone obtains
a reference to a frame (for example through the
sd_lldp_get_tlv_packets() function), the frame is not destroyed
immediately upon removal from MIB; it gets destroyed only when the
last user releases it.

**Comment by bengal** [comment]
Any news on this?

**Comment by bengal** [comment]
@ssahani, @teg, @poettering, could you please give some feedback on this?

**Comment by dvdhrm** [comment]
So this looks mostly good. However, I'd _really_ prefer if someone writes a test that actually runs the lldp state-machine (requests can be served statically). It is a mess to test this right now, and I'd like to avoid merging further lldp patches without a proper test-suite for it.

If I'm not mistaken, it should be easy to run this over a virtual link and server the lldp requests with a static hard-coded reply?

File: src/systemd/sd-lldp.h [hunk 1]
```diff
@@ -25,6 +25,7 @@
 #include "sd-event.h"
 
 typedef struct sd_lldp sd_lldp;
+typedef struct tlv_packet sd_lldp_tlv_packet;
```
Comment (dvdhrm) [inline src/systemd/sd-lldp.h:28]: This looks wrong. This should be `typedef struct sd_lldp_tlv_packet sd_lldp_tlv_packet;`

File: src/systemd/sd-lldp.h [hunk 2]
```diff
@@ -40,6 +41,12 @@ typedef enum LLDPPortStatus {
         _LLDP_PORT_STATUS_INVALID = -1,
 } LLDPPortStatus;
 
+typedef enum LLDPDestinationType {
+        LLDP_DEST_TYPE_NEAREST_BRIDGE,
+        LLDP_DEST_TYPE_NEAREST_NON_TPMR_BRIDGE,
+        LLDP_DEST_TYPE_NEAREST_CUSTOMER_BRIDGE,
+} LLDPDestinationType;
```
Comment (dvdhrm) [inline src/systemd/sd-lldp.h:48]: This _must_ be namespaced properly. Please prefix with `SD_LLDP_DESTINATION_*`. Also please avoid abbreviations.

Note that I just pushed some cleanup-patches to -git which fix the other name-spacing issues.

File: src/systemd/sd-lldp.h [hunk 3]
```diff
@@ -51,3 +58,25 @@ int sd_lldp_detach_event(sd_lldp *lldp);
...
+int sd_lldp_tlv_packet_read_vlan_name(sd_lldp_tlv_packet *tlv, uint16_t *vlan_id, char **name, uint16_t *length);
+int sd_lldp_tlv_packet_read_management_vid(sd_lldp_tlv_packet *tlv, uint16_t *id);
+int sd_lldp_tlv_packet_read_link_aggregation(sd_lldp_tlv_packet *tlv, uint8_t *status, uint32_t *id);
+
+int sd_lldp_tlv_packet_get_destination_type(sd_lldp_tlv_packet *tlv, LLDPDestinationType *dest);
+
+sd_lldp_tlv_packet *sd_lldp_tlv_packet_ref(sd_lldp_tlv_packet *tlv);
+sd_lldp_tlv_packet *sd_lldp_tlv_packet_unref(sd_lldp_tlv_packet *tlv);
+
+int sd_lldp_get_tlv_packets(sd_lldp *lldp, sd_lldp_tlv_packet ***tlvs, unsigned max);
```
Comment (dvdhrm) [inline src/systemd/sd-lldp.h:82]: Why the "max" argument?

File: src/libsystemd-network/lldp-tlv.h [hunk 4]
```diff
@@ -76,6 +87,7 @@ int tlv_packet_append_u32(tlv_packet *m, uint32_t data);
 int tlv_packet_append_string(tlv_packet *m, char *data, uint16_t size);
 
 int lldp_tlv_packet_enter_container(tlv_packet *m, uint16_t type);
+int lldp_tlv_packet_enter_container_org(tlv_packet *m, const uint8_t *oui, uint8_t subtype);
```
Comment (dvdhrm) [inline src/libsystemd-network/lldp-tlv.h:90]: "org"? Please no abbreviation, except for _local_ variable declarations.

**Comment by dvdhrm** [comment]
Also: Why is this called "tlv_packet" instead of just "packet"? I'd prefer if we rename it to "sd_lldp_packet" and "lldp_packet_section".

If we need a generic TLV parser, we can add it to src/basic. But in this context, I cannot see why we need the "tlv" in the object names. @ssahani, any comment?

**Comment by bengal** [comment]
> So this looks mostly good. However, I'd really prefer if someone writes a test that actually runs the lldp state-machine (requests can be served statically). It is a mess to test this right now, and I'd like to avoid merging further lldp patches without a proper test-suite for it.
> 
> If I'm not mistaken, it should be easy to run this over a virtual link and server the lldp requests with a static hard-coded reply?

Yes, that would be simple to implement, but at the moment there isn't
any way to obtain the neighbor MIB using public API. Therefore in my
opinion it makes sense to add such test cases in the next PR after
other commits, so that it will be possible to implement test cases
like:
1. create socket pair
2. run LLDP instance attached to one socket
3. send hardcoded message(s) to other socket
4. get list of neighbors with sd_lldp_get_tlv_packets()
5. verify that neighbors list contains the expected entries

What do you think?

File: src/systemd/sd-lldp.h (hunk 3 above)
Comment (bengal) [inline src/systemd/sd-lldp.h:82]: > Why the "max" argument?

To limit the number of returned elements, or (unsigned) -1 to return all of them -- see https://github.com/systemd/systemd/pull/486#commitcomment-12094189.

File: src/libsystemd-network/lldp-tlv.h (hunk 4 above)
Comment (bengal) [inline src/libsystemd-network/lldp-tlv.h:90]: > "org"? Please no abbreviation, except for local variable declarations.

Well, lldp_tlv_packet_enter_organizationally_specific_container() seemed too long. How about one of the following?

 lldp_tlv_packet_enter_container_oui()
 lldp_tlv_packet_enter_container_oui_subtype()

Comment (dvdhrm) [inline src/libsystemd-network/lldp-tlv.h:90]: OUI sounds much better.

File: src/systemd/sd-lldp.h (hunk 3 above)
Comment (dvdhrm) [inline src/systemd/sd-lldp.h:82]: Why would anyone pass something here? Yeah, I can see the comment by Lennart, but I cannot see a justification why that is needed? Just make it return the full TLV list at all times?

**Comment by dvdhrm** [comment]
Can you include such test in this PR? I really feel uncomfortable merging more LLDP patches without any runtime test. Reviewing the code, I saw lots of issues in the LLDP stream-parser, which I cannot verify without a proper test.

In general, I'm fine with this series.

**Comment by ssahani** [comment]
@dvdhrm the reason behind putting a tlv is LLDP packets encoded as tlv https://en.wikipedia.org/wiki/Type-length-value.

This TLV parser/encoder are specifically designed for LLDP. I am not sure whether if any other protocols can be parsed or encoded by it .

**Comment by ssahani** [comment]
@bengal I  get it what this patch is doing.  But you missing the point is we need to drop the packets 
as soon as TTL expires and the packet processing engine should do it.
Now the expiry logic thinks that the packets is dropped but infact is not.  The packet is still there.

We have limited the number of packets  to LLDP_MIB_MAX_PORT_PER_CHASSIS .
Now if you take a ref the limit will be crossed . If we never unref the packet the system will run out of memory (OOM). 
I am fine with taking a copy of the packet via DBUS or some other IPC .

**Comment by bengal** [comment]
> the packet we receive it always has the data already. We try mostly to avoid duplicating and storing. It always a consideration of CPU vs Mem . Can this be avoided ?

Yes, a pointer to frame data can be used instead.

**Comment by bengal** [comment]
> Now the expiry logic thinks that the packets is dropped but infact is not. The packet is still there.
> We have limited the number of packets to LLDP_MIB_MAX_PORT_PER_CHASSIS .
> Now if you take a ref the limit will be crossed . 

Yes, the idea is that the packet is shared among different users and
the last one unreferencing it causes its destruction.

But this is totally transparent to the LLDP state machine because when the
TTL for a packet expires lldp_neighbour_port_remove_and_free() unrefs
it and then decreases chassis->n_ref. At this point the LLDP MIB
doesn't contain the packet anymore and there isn't any risk of
exceeding LLDP_MIB_MAX_PORT_PER_CHASSIS.

> If we never unref the packet the system will run out of memory (OOM).

The packet data is exposed through an API which dictates that the
packet must be unreferenced when no more in use, if this is not done
it would be a programming error by the caller.

> I am fine with taking a copy of the packet via DBUS or some other IPC .

OK, this can be an alternative, but I think it would be more
efficient if we can avoid to copy every packet.

**Comment by ssahani** [comment]
Isn't what exactly what I pointed out . The packets is still there and lib thinks it's gone so we are crossing
the limits by storing more packets which actually violates the whole idea of limiting packets.
The third party user not sure when it will release the packet. By the time if we flood with LLDP packets
we will get into OOM. The library is tested with stress test by flooding LLDP packets.

**Comment by bengal** [comment]
> Isn't what exactly what I pointed out . The packets is still there and lib thinks it's gone so we are crossing
> the limits by storing more packets which actually violates the whole idea of limiting packets.
> The third party user not sure when it will release the packet. By the time if we flood with LLDP packets
> we will get into OOM. The library is tested with stress test by flooding LLDP packets.

I understand your concerns about memory usage, but in my opinion this
is not something that should be enforced by the API; once
get_lldp_packets() returns, the caller (partially) owns the packets
and is responsible for their memory. This is a library API and users
not releasing packets are simply misusing the library.

The typical usage of this API will be an application starting LLDP on
an interface and periodically obtaining the neighbor lists. So at
every time there will be at most 32K (MIB) + 32K (referenced by user)
packet entries allocated (in the worst case).

Do you have any other approach to suggest?

**Comment by ssahani** [comment]
Usually we do DBUS approach.  I would request @poettering , @teg  to see what approach we take?

**Comment by ssahani** [comment]
BTW The reason behind limiting the packets is  for security . If someone try to exploit the system by flooding with LLDP packets the system should not goes down.  The library should able to handle this scenario and act on it's own, Rather than storing whatever is coming over.

For example we modify the TTL of LLDP packets to for example 10000 or more and keep on flooding packets. After sometime if there is no limits it will get into OOM.

**Comment by bengal** [comment]
> Usually we do DBUS approach. I would request @poettering , @teg to
> see what approach we take?

Note that here we are talking about how to make available LLDP
neighbors through the libsystemd-network library, not to external
applications. My understanding is that libsystemd-network will become
one day (soon?) an independent library for applications to link
against. In NetworkManager we are already reusing some of the library
code for DHCP, IPv4LL and hopefully will do the same also for LLDP.

**Comment by bengal** [comment]
> For example we modify the TTL of LLDP packets to for example 10000 or more and keep on flooding packets. After sometime if there is no limits it will get into OOM.

As explained above, the amount of stored packets will still have an
upper bound.

I'm going to open a new pull request to address the other comments
above, we can continue there the discussion regarding alternative
approaches.

**Comment by bengal** [comment]
Closing, refer to #1448 for the new version.

**Comment by ssahani** [comment]
for accessing the LLDP packets we have already provision.
sd_lldp_set_callback

Please looks at src/network/networkd-link.c how we set it .
In this way you can register your CB and you get the same packet as done in lldp_handler. 
Now we do sd_lldp_save for networkctl .

**Pull request closed without merging**